### PR TITLE
Feature/token purge

### PR DESCRIPTION
**Pull request by pliesveld**
Added a scheduled task to remove accounts that have not been enabled and have an expired verification token.

### CODE REVIEW

File: src/main/java/org/baeldung/test/TestDBConfig.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
+package org.baeldung.test;
```
Comment (eugenp) [inline src/main/java/org/baeldung/test/TestDBConfig.java]: Since this is testing only - this should be under /src/test/java

**Comment by eugenp** [comment]
Thanks for the PR - looks good (and quite unexpected). 
I only made a small comment about testing artifacts being stored in /src/test/java (that one and any others that are only used for testing) and it looks ready to go (I haven't run it yet). 
And of course, since you went through the trouble of getting it implemented, it'd be happy to publish a writeup about it if you want to give that a go. 
Cheers, 
Eugen.

**Comment by pliesveld** [comment]
Thanks for the feedback, and your articles. I saw two spring test configs, and choose poorly.

I noticed that the purge task didn't check for Password Reset tokens, and modified it accordingly.  But I have a few comments.

Would it make more sense to have the purge task call UserService#deleteUser() ?

Right now; the purge task loads all expired tokens into memory before removing users.  Should I handle this operation in batches?

Would it be logical to change entity associations from uni-directional to bi-directional, and then utilize _CascadeType.REMOVE_ and _orphanRemoval_ from the parent entity?

In the project that I am working on when I followed your tutorials, instead of having a child entity with a surrogate key, I used a shared primary key with a hibernate specific operation to add cascade delete to the generated DDL.  Something like the following:



    @Entity
    public class ChildEntity
    {
        @Id
        @Column(name = "ENTITY_ID")
        private Integer id;

        @MapsId
        @OneToOne(cascade = CascadeType.ALL)
        @JoinColumn(name = "ENTITY_ID")
        private Parent parent;
    }

    @Entity
    public class Parent
    {
        @Id
        @GeneratedValue(strategy = GenerationType.AUTO)
        @Column(name = "ENTITY_ID")
        private Integer id;

        @OneToOne(cascade = CascadeType.ALL, orphanRemoval = true)
        @PrimaryKeyJoinColumn
        @org.hibernate.annotations.OnDelete(action = org.hibernate.annotations.OnDeleteAction.CASCADE)
        private ChildEntity childEntity;
    }

Would such an approach have any benefit here?

Lastly; I would be happy to do a writeup.

-patrick

**Comment by eugenp** [comment]
Hey Patrick - that's an interesting point. 
My thinking for the general-purpose case is that you should actually only clear the expired tokens, but leave the disabled users in the DB. 
Ultimately it's the token that expires and should be removed. The user may have simply forgotten to verify their email in time - in which case, it's probably easier to build a basic mechanism that allows them to re-trigger the email verification than to go through the process all over again. 
Keep in mind that this will depend on your usecase, so it may be perfectly valid to remove them if that's the business requirement. And in that case - yes, it makes perfect sense to have a relation between the entities and cascade the delete. 
The POC has no external business requirement, so my goal is to go with whatever makes more sense. 
Finally, as for the writeup - that's definitely going to be interesting. Ping me over email or via the contact for of the site (www.baeldung.com/contact) and I'll send you the onboarding info about that. 
Cheers, 
Eugen.

**Pull request merged**